sdk/go/common/util/env/env.go

```diff
@@ -0,0 +1,200 @@
+// Copyright 2016-2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// A small library for creating consistent and documented environmental variable accesses.
+//
+// Public environmental variables should be declared as a module level variable.
+package env
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+)
+
+var Experimental = Bool("EXPERIMENTAL", "Enable experimental options and commands")
```

Comment:
Ah! I see that enviroment variables aren't necessarily declared in the library but can occur anywhere in the program.

I would encourage us to have one file that colocates all environment variable declarations. This makes it a lot easier to see what the available configuration options are, and to know immediately where to look when you need a particular env var.

I suppose my suggestion above of a `Environment` struct would necessitate that design.

Reply:
Yes, environmental variables can be declared anywhere. My assumption is that they will be declared in the module where they are consumed.

I disagree with this. I think that environmental variables should be declared in the module where they are used. We can let the runtime aggregate all declared variables on our behalf. If we want a list of existing env vars, we can call `pulumi env`.

Reply:
IMO I don't believe there's enough benefit to colocating the configurable with the location they're used. It's better to declare them in one location, then perform any validation/munging necessary before initializing the final `Configuration` struct which is known to be well-formed. This allows you to "fail fast" -- return an error to the user before any real computation is done. 

It would be nice if we can put all validation logic for configurables in a single place. That way, deeply nested code isn't trying to check the validity of configuration.

For example, `--parallel=-1` should be checked when the program starts, not when the engine is creating a work queue for the `RegisterResource` requests. Colocating all the logic in place makes that possible (practicing RAII for config).